### PR TITLE
#54 backend improvments of attraction domain

### LINKED ISSUE #54: Improvements of Attraction domain
# Attraction Domain Model Enhancements

## Overview
This document outlines the proposed enhancements to improve the Attraction domain model, addressing specific business requirements.

#### Description:
- **Functionality**: An API endpoint to search for attractions based on query parameters such as country, state, and city.
- **Implementation**: A GET request structured as `/attractions?country={country}&state={state}&city={city}`. This endpoint will dynamically handle these parameters to filter and return relevant attraction data.
- **get_cities endpoint**: We need to perform endpoint which will retrieve cities for pointed state

#### Description:
- **Purpose**: To calculate a radius based on a user-specified location and distance.
- **Implementation**: A method that takes a point (latitude and longitude) and a distance (in kilometers or miles) to calculate the circular area (radius) around the point, using the formula adapted to geographical coordinates. This method is essential for identifying attractions within a given distance from the user's chosen point.

#### Description:
- **Objective**: To develop an algorithm that aggregates likes and scores for each attraction, ranking them based on popularity and ratings.
- **Categories**:
  - *Top Rated*: Attractions with the highest average scores.
  - *Most Rated*: Attractions with the highest number of ratings or likes.
- **Implementation Considerations**: This feature will require a robust data aggregation and sorting mechanism. Periodic caching of results may be beneficial for optimizing performance.

#### Description
- **Objective**: There is need to add additional fields to model.
 - * Category of Attraction: We need to set category, which will be choice filed with values (cultural, sports)

### PR DESCRIPTION
**Pull request by kwiats**

### CODE REVIEW

File: backend/src/attraction/interfaces/repository.py [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
+
 
 class Repository(ABC):
     @abstractmethod
     def get_all(self, db):
         pass
 
+    @abstractmethod
+    def get_filtered(self, db, filters: AttractionFilters):
+        pass
```
Comment (KarolBautrel) [inline backend/src/attraction/interfaces/repository.py]: maybe get_by_filters?

File: backend/src/attraction/repositories/attraction_repo.py [hunk 2]
```diff
@@ -14,6 +21,45 @@ class AttractionRepository(Repository):
...
+            ).cast(Geography)
+            attraction_point = func.ST_SetSRID(
+                func.ST_MakePoint(Attraction.longitude, Attraction.latitude), 4326
+            ).cast(Geography)
+            query = query.filter(
+                func.ST_DWithin(attraction_point, point, filters.radius)
+            )
+
+        # sorting
+        if filters.sort_by == AttractionSortedBy.TopRated:
```
Comment (KarolBautrel) [inline backend/src/attraction/repositories/attraction_repo.py]: Maybe we can do this in better way ? For example with use of dictionary ? for example 
```

    filter_operations = {
        'city': lambda value: Attraction.city == value,
        'country': lambda value: Attraction.country == value,
        'category': lambda value: Attraction.category == value,
    }

   
    for filter_name, operation in filter_operations.items():
        filter_value = getattr(filters, filter_name, None)
        if filter_value:
            query = query.filter(operation(filter_value))
```

File: backend/src/attraction/routes/router.py [hunk 3]
```diff
@@ -22,6 +22,15 @@ def get_all_attractions(
     return _attraction_service.get_all_attractions(db)
 
 
+@router.get("")
+def get_attractions(
+    filters: AttractionFilters = Depends(),
```
Comment (KarolBautrel) [inline backend/src/attraction/routes/router.py:27]: Is that queryparams ?

File: backend/src/core/common_schema.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    page: int = 1
+    per_page: int = 25
+
+    @property
+    def offset(self) -> int:
+        return (self.page - 1) * self.per_page
+
+    @property
+    def limit(self) -> int:
+        return self.per_page
```
Comment (KarolBautrel) [inline backend/src/core/common_schema.py:16]: Im not sure if property works with pydantic

File: backend/src/attraction/routes/router.py (hunk 3 above)
Comment (kwiats) [reply backend/src/attraction/routes/router.py:27]: Yes, this is query params

File: backend/src/core/common_schema.py (hunk 4 above)
Comment (kwiats) [reply backend/src/core/common_schema.py:16]: Works, i tested this and it's fine

File: backend/src/attraction/routes/router.py (hunk 3 above)
Comment (KarolBautrel) [reply backend/src/attraction/routes/router.py:27]: Shouldnt we use it like this ? 
https://fastapi.tiangolo.com/tutorial/query-params/

**Pull request merged**